### PR TITLE
AWS plugin update to support additional metadata in responseTransformation

### PR DESCRIPTION
**Pull request by jbohanon**
# Description

AWS plugin updates that will allow the `responseTransformation` feature to directly use the unwrapAsApiGateway transformer.

# Context

Users have requested this feature to allow for transformation of lambda response bodies using the `responseTransformation` feature.

# Testing
This feature can be manually tested by building the envoy-gloo wrapper with the command `ENVOY_GLOO_IMAGE=gcr.io/solo-test-236622/envoy-gloo:4438-response-transformation make push-kind-images build-test-chart` and deploying according to the [guide](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/aws_lambda/).

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 1]
```diff
@@ -285,6 +285,9 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
 									Text: "text/html",
 								},
 							},
+							HeadersExtractionKey: "headers",
+							StatusExtractionKey:  "statusCode",
+							BodyExtractionKey:    "body",
```
Comment (ben-taussig-solo) [inline projects/gloo/pkg/plugins/aws/plugin.go:290]: Doesn't this line do the same thing as 278-282?

Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:290]: Yes thanks I meant to remove that

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 2]
```diff
@@ -285,6 +285,9 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
 									Text: "text/html",
```
Comment (jbohanon) [inline projects/gloo/pkg/plugins/aws/plugin.go:285]: @ben-taussig-solo how do you feel about introducing an option to disable this part? Seems a bit outdated but some customers might rely on the functionality

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 3]
```diff
@@ -236,31 +239,24 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
...
 
-			requiresUnwrap := awsDestinationSpec.Aws.GetUnwrapAsAlb() || awsDestinationSpec.Aws.GetUnwrapAsApiGateway()
-			if requiresUnwrap {
-				// if we are unwrapping the response and have no request transformation, we can return early
-				if !requiresRequestTransformation {
-					return nil, nil
-				}
-				// Do not transform response if we are unwrapping since response is handled in its entirety by these
-				requiresResponseTransformation = false
+			// We should have already mutated awsDestination.UnwrapAsApiGateway if awsDestination.ResponseTransformation
```
Comment (ben-taussig-solo) [inline projects/gloo/pkg/plugins/aws/plugin.go:246]: where would this occur? in solo-projects?

Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/aws/plugin.go:246]: you literally explain it on the next line....

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 4]
```diff
@@ -236,31 +239,24 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
...
+			// is set to true. This is done in GenerateAWSLambdaRouteConfig
+			requiresUnwrap :=
+				// unwrapAsAlb is handled directly in the aws lambda filter rather than through a transformer.
```
Comment (ben-taussig-solo) [inline projects/gloo/pkg/plugins/aws/plugin.go:249]: Is this true?

Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/aws/plugin.go:249]: sorry, confused this with unwrapAsApiGateway

**Comment by ben-taussig-solo** [comment]
Going to spend a little more time parsing through the lambda filter logic. I think the most important thing here is to keep in mind that there's a lot of interaction b/w the enterprise lambda plugin and the open source lambda plugin we're modifying here. So we should be careful not to merge this until we're sure that lambda behaves as expected in both enterprise and open source.

Also -- Do we need to wait for an Envoy version bump for this PR to be ready to merge? If so could we put the WIP tag on this PR?

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 5]
```diff
@@ -236,31 +239,24 @@ func (p *Plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
...
+				// If both are set, we prioritize unwrapAsAlb.
```
Comment (ben-taussig-solo) [inline projects/gloo/pkg/plugins/aws/plugin.go:250]: Could you add a test against this?

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 3 above)
Comment (ben-taussig-solo) [inline projects/gloo/pkg/plugins/aws/plugin.go:246]: If possible, could we test against this?

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 5 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:250]: added in fdf8f72b661d15272aa58aad968fab435c5304b7

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 3 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:246]: added in fdf8f72b661d15272aa58aad968fab435c5304b7

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 5 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/aws/plugin.go:250]: Looks great!

**Comment by jbohanon** [comment]
/kick [FAIL] Consul + Vault Configuration Happy Path e2e [BeforeEach] can do function routing with consul services

File: changelog/v1.14.0-beta15/merge-responsetransformation-unwrapas.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+changelog:
+- type: NON_USER_FACING
+  issueLink: https://github.com/solo-io/solo-projects/issues/4438
+  resolvesIssue: false
+  description: >-
+    Bring responseTransformation to parity with unwrapAsApiGateway in lambda filter
+- type: NEW_FEATURE
```
Comment (sam-heilbron) [inline changelog/v1.14.0-beta15/merge-responsetransformation-unwrapas.yaml:7]: In our changelog, we have a lot of useful content (new feature, breaking change) that references an enterprise changelog (which won't be visible to users). Can we create an open source issue to track this work broadly so that our changelog has links that users can see? I figure especially since we're making more of these features available in open source this would be helpful

File: cloudbuild.yaml [hunk 7]
```diff
@@ -41,7 +41,9 @@ steps:
   - '-c'
   - |
     ENVOY_VERSION=v$$(make print-ENVOY_GLOO_IMAGE | cut -d: -f2)
-    gsutil cp gs://solo-public-artifacts.solo.io/envoy/$$ENVOY_VERSION/envoy.stripped /workspace/envoy
+    # DO NOT MERGE without reverting
```
Comment (sam-heilbron) [inline cloudbuild.yaml:44]: Out of curiosity, why is this required instead of just setting "ENVOY_GLOO_IMAGE" to the SHA of your build? Is it because it's in a different registry?

File: projects/gloo/api/v1/options/aws/aws.proto [hunk 8]
```diff
@@ -74,29 +74,31 @@ message DestinationSpec {
         SYNC = 0;
         ASYNC = 1;
     }
-    // Include headers, querystring, request path, and request method in the event payload sent to aws lambda
+
+    // Include headers, querystring, request path, and request method in the event payload sent to aws lambda.
+    // Only one of `requestTransformation` or `wrapAsApiGateway` should be provided.
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/options/aws/aws.proto:79]: Why is request_transformation not deprecated like the response_transformation flag?

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 9]
```diff
@@ -413,17 +389,30 @@ func GenerateAWSLambdaRouteConfig(options *v1.GlooOptions_AWSOptions, destinatio
 			awsRegion, upstream.GetAwsAccountId(), functionName)
 	}
 
+	// transparently use the unwrapAsApiGateway functionality for responseTransformation
+	if destination.GetResponseTransformation() {
+		destination.UnwrapAsApiGateway = true
+	}
+
+	var transformerConfig *v3.TypedExtensionConfig
+	if !destination.GetUnwrapAsAlb() && destination.GetUnwrapAsApiGateway() {
```
Comment (sam-heilbron) [inline projects/gloo/pkg/plugins/aws/plugin.go:398]: Perhaps this is just a remnant of the API evoltion, but I'm a little confused about how 
- UpwrapAsAlb
- UpwrapAsApiGateway
- Transformers 

play together. We explicitly pass UnwrapAsAlb in the LambdaPerRoute object, there is no UnwrapAsApiGateway reference. Is the plan to eventually remove that field as well and everything flows through the transformer config?

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 5 above)
Comment (sam-heilbron) [inline projects/gloo/pkg/plugins/aws/plugin.go:250]: Is this a new behavior (if both are set we prioritize one)? Did you consider adding an error instead? I'm not tied to either direction, more curious if it was considered

**Review by sam-heilbron — COMMENTED** [review]
Looking great! Small comments/questions

File: projects/gloo/api/v1/options/aws/aws.proto (hunk 8 above)
Comment (jbohanon) [reply projects/gloo/api/v1/options/aws/aws.proto:79]: because `requestTransformation` is still substantively different from `wrapAsApiGateway` and there still exists a use case for each. It makes more sense to deprecate `responseTransformation` than `unwrapAsApiGateway` since the latter is more specific, despite the fact that they will now be identical

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 9 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:398]: Yes, it would be ideal for unwrapAsAlb to be moved into its own transformer ([issue link](https://github.com/solo-io/envoy-gloo/issues/205)) to clean up this whole ordeal. UnwrapAsAlb is already listed as deprecated in our internal API, not sure why it was not marked in the public API. I meant to ask but forgot

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 5 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:250]: This is not new behavior. I think it would be cleaner to error instead, and I'm happy to make that change here if it wouldn't be considered too breaking

File: cloudbuild.yaml (hunk 7 above)
Comment (jbohanon) [reply cloudbuild.yaml:44]: It all worked locally, but I wasn't able to control that env var for CI (unless I missed something) so I was trying to mess with this to get inter-PR CI to work

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 9 above)
Comment (sam-heilbron) [reply projects/gloo/pkg/plugins/aws/plugin.go:398]: Yeah so since it's deprecated, can we try to get to the bottom of if we even need to set it? If we have to keep it around and there is a follow up issue to clean it up, I can live with that as well

File: projects/gloo/api/v1/options/aws/aws.proto (hunk 8 above)
Comment (sam-heilbron) [reply projects/gloo/api/v1/options/aws/aws.proto:79]: ah makes sense, thanks!

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 5 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:250]: Per offline discussion, this will be changed to error instead of applying priority

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 9 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:398]: We need to keep unwrapAsAlb functionality, but the open issue linked above will be the condition where this option can be removed/reserved

File: changelog/v1.14.0-beta15/merge-responsetransformation-unwrapas.yaml (hunk 6 above)
Comment (jbohanon) [reply changelog/v1.14.0-beta15/merge-responsetransformation-unwrapas.yaml]: https://github.com/solo-io/gloo/issues/7907

**Comment by jbohanon** [comment]
/kick https://github.com/solo-io/gloo/issues/7932 
<details>
<summary>logs</summary>

Step 8 - "run-e2e-tests": Summarizing 1 Failure:
Step 8 - "run-e2e-tests":   [FAIL] Consul e2e with envoy and gloo test eds only updates blocking EDS queries [It] works as expected
Step 8 - "run-e2e-tests":   /workspace/gloo/test/e2e/consul_test.go:293
</details>

File: changelog/v1.14.0-beta16/merge-responsetransformation-unwrapas.yaml
Comment (jbohanon) [inline changelog/v1.14.0-beta16/merge-responsetransformation-unwrapas.yaml:1]: add dep bump changelog entry

File: projects/gloo/pkg/plugins/enterprise_warning/plugin_test.go
Comment (jbohanon) [inline projects/gloo/pkg/plugins/enterprise_warning/plugin_test.go:1]: these tests should be changed to check for aws.WrapAsApiGateway instead of being removed

File: Makefile [hunk 10]
```diff
@@ -89,7 +89,7 @@ else
   endif
 endif
 
-ENVOY_GLOO_IMAGE ?= quay.io/solo-io/envoy-gloo:1.24.0-patch2
+ENVOY_GLOO_IMAGE ?= quay.io/solo-io/envoy-gloo:1.25.1-patch2
```
Comment (sam-heilbron) [inline Makefile:92]: How does this upgrade affect the inflight work outlined here: https://github.com/solo-io/solo-projects/issues/4462?

Are there any important changes from 1.24->1.25 that we should callout to users or improved APIs that we should be using? (I dont think we need to do that as part of this PR, I just wonder if there are issues we can create to benefit from new features)

File: test/e2e/aws_test.go [hunk 11]
```diff
@@ -149,10 +149,7 @@ var _ = Describe("AWS Lambda", func() {
 		}))
 	}
 
-	testProxy := func() {
-		err := envoyInstance.RunWithRoleAndRestXds(services.DefaultProxyName, testClients.GlooPort, testClients.RestXdsPort)
-		Expect(err).NotTo(HaveOccurred())
-
+	createProxy := func(unwrapAsApiGateway, requestTransformation, responseTransformation bool, logicalName string) {
```
Comment (sam-heilbron) [inline test/e2e/aws_test.go:152]: I know you're working within the existing test framework, so feel free to ignore/punt on the following proposal:

It feels like these tests are well suited for a table driven test. We have a number of different API's that overlap, and varying behavior that we want to support. What if we updated this test suite to use a table format so that its eeasy to follow what the behaviors are and easy to add when we need to handle new cases.

Again, I think we can certainly do this after this work merges, but since there have been a number of lambda related questions and its a space we plan to expand, having strong testing will make our lives easier in the future

**Review by sam-heilbron — COMMENTED** [review]
CHanges look good to me! Some small questions

File: test/e2e/aws_test.go (hunk 11 above)
Comment (jbohanon) [reply test/e2e/aws_test.go:152]: https://github.com/solo-io/gloo/issues/7970

File: Makefile (hunk 10 above)
Comment (jbohanon) [reply Makefile:92]: @nfuden Do you have more context on the issue Sam linked? Do we have breakage involved in the 1.25 update that needs to be addressed before we bump?

File: projects/gloo/api/v1/options/aws/aws.proto [hunk 12]
```diff
@@ -74,29 +74,29 @@ message DestinationSpec {
...
     bool request_transformation = 6;
-    
-    // de-jsonify response bodies returned from aws lambda
-    bool response_transformation = 5;
+
+    // Deprecated. Use unwrapAsApiGateway
```
Comment (ben-taussig-solo) [inline projects/gloo/api/v1/options/aws/aws.proto:82]: Is this actually deprecated if we're making updates to the setting in this PR? I know there was some discussion about this a couple of weeks ago, could you catch me up to speed?

Comment (jbohanon) [reply projects/gloo/api/v1/options/aws/aws.proto:82]: The plan as of now is to deprecate the more generic `responseTransformation` in favor of more specific transformations a la `unwrapAsApiGateway` and `unwrapAsAlb` in addition to any others that may come along in the future

File: Makefile (hunk 10 above)
Comment (nfuden) [reply Makefile:92]: We should be good we have fixed most of them we still need to do some work on the caching updates provided by upstream but we fixed all the in envoy issues and the go-control-plane bump didnt seem to break things. I ran through some extra manual testing on this branch and we should be good to go with regards to the upgrade of envoy

**Comment by jbohanon** [comment]
> Leaving this in WIP until upgrade issues with envoy 1.25 are resolved.
> 
> 
> 
> https://github.com/solo-io/solo-projects/issues/4462

Resolved per Nathan's comment

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 13]
```diff
@@ -413,17 +389,35 @@ func GenerateAWSLambdaRouteConfig(options *v1.GlooOptions_AWSOptions, destinatio
 			awsRegion, upstream.GetAwsAccountId(), functionName)
 	}
 
+	// transparently use the unwrapAsApiGateway functionality for responseTransformation
+	if destination.GetResponseTransformation() {
```
Comment (sam-heilbron) [inline projects/gloo/pkg/plugins/aws/plugin.go:393]: Since our API is marked as deprecated, I think it would be valuable if we had a way of alerting users that they are using deprecated functionality. In an ideal world, this would be handled in the proto, but since that does not occur, what do you think about a log message warning about configuration that uses deprecated (responseTransformation) functionality?

**Review by sam-heilbron — COMMENTED** [review]
The changes look great! 

How do we plan on documenting the behavior change that we now return an error when both UnwrapAsAlb and UnwrapAsApiGateway are set? Is this something we can do in a changelog or are you just planning on adding it to the upgrade documentation for 1.14?

I think we can flush out the e2e tests a bit more. We claim that we are introducing certain behaviors:
- responseTransformation will no longer overwrite the content-type header to text/html.
- responseTransformation will now extract headers and status code from response body in addition to the body.
- Bring responseTransformation to parity with unwrapAsApiGateway in lambda filter

But I don't see tests to validate these behaviors. Am I missing something?

**Comment by jbohanon** [comment]
> How do we plan on documenting the behavior change that we now return an error when both UnwrapAsAlb and UnwrapAsApiGateway are set? Is this something we can do in a changelog or are you just planning on adding it to the upgrade documentation for 1.14?

I can add a `BREAKING_CHANGE` changelog entry. I think it would also be valuable to add to the upgrade documentation.

> I think we can flush out the e2e tests a bit more. We claim that we are introducing certain behaviors:
> 
> * responseTransformation will no longer overwrite the content-type header to text/html.
> * responseTransformation will now extract headers and status code from response body in addition to the body.
> * Bring responseTransformation to parity with unwrapAsApiGateway in lambda filter
> 
> But I don't see tests to validate these behaviors. Am I missing something?

The first bullet is a consequence of deleted code moreso than any additive changes. I think it could be confusing to future devs to test that removed functionality is not there. The second bullet could certainly use more discrete testing in the e2e context. The lattermost bullet here is just a summary of the first two.

File: projects/gloo/pkg/plugins/aws/plugin.go (hunk 13 above)
Comment (jbohanon) [reply projects/gloo/pkg/plugins/aws/plugin.go:399]: I have added the warning log here: 81b3d6c88b76bb55ef844a31213198ff408522c7

File: docs/content/operations/upgrading/v1.14.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
+---
```
Comment (sam-heilbron) [inline docs/content/operations/upgrading/v1.14.md:1]: This is awesome! I think having this in the repo early will be helpful to ensure we encourage all developers who introduce changes that affect UX to outline the impact here

**Review by sam-heilbron — COMMENTED** [review]
Loooking great!

one nit on they way that the e2e tests assert behaviors of the http response. Not a blocker, but I think worth a discussion to try to consolidate our approach to writing http response matchers.

File: test/e2e/aws_test.go [hunk 15]
```diff
@@ -84,34 +87,130 @@ var _ = Describe("AWS Lambda", func() {
...
+		if params.requestUrl != nil {
+			u = params.requestUrl
+		}
+		expectedStatus := http.StatusOK
+		if params.expectedStatus != nil {
+			expectedStatus = *params.expectedStatus
+		}
 
-		EventuallyWithOffset(offset, func() (string, error) {
+		EventuallyWithOffset(params.offset, func() error {
```
[the remark this answers is not in the document]
Comment (jbohanon) [reply test/e2e/aws_test.go]: Updated to use standard matcher [8a7acdf](https://github.com/solo-io/gloo/pull/7854/commits/8a7acdf5bedffd203870893747bea7c721e752e1)

File: test/gomega/matchers/http.go [hunk 16]
```diff
@@ -58,6 +60,26 @@ func HaveOkResponseWithHeaders(headers map[string]interface{}) types.GomegaMatch
...
+		vals := make([]interface{}, len(v))
+		for i := range v {
+			vals[i] = v[i]
+		}
+		headerMatchers = append(headerMatchers, gomega.WithTransform(transforms.WithHeaderValues(k), gomega.ContainElements(vals...)))
+	}
+	return gomega.And(headerMatchers...)
+}
+
+func MatchMultipleSubstrings(substrings []string) types.GomegaMatcher {
```
Comment (sam-heilbron) [inline test/gomega/matchers/http.go:75]: nit: Could we name this ContainSubstrings? That follows the naming convention of the base GomegaMatcher 'ContainSubstring' since we're building a wrapper around that for multiple values.

Also can we add a comment to his exported function to define the behavior

File: test/gomega/matchers/http.go [hunk 17]
```diff
@@ -58,6 +60,26 @@ func HaveOkResponseWithHeaders(headers map[string]interface{}) types.GomegaMatch
 	})
 }
 
+func MatchMultiValueHeaders(headers http.Header) types.GomegaMatcher {
```
Comment (sam-heilbron) [inline test/gomega/matchers/http.go:63]: nit: ContainHeaders?

File: test/gomega/transforms/transforms.go [hunk 18]
```diff
@@ -20,7 +20,7 @@ func WithDecompressorTransform() func(b []byte) string {
 			return invalidDecompressorResponse
 		}
 		defer reader.Close()
-		body, err := ioutil.ReadAll(reader)
+		body, err := io.ReadAll(reader)
```
Comment (sam-heilbron) [inline test/gomega/transforms/transforms.go:23]: 👏

File: test/e2e/aws_test.go [hunk 19]
```diff
@@ -84,34 +87,69 @@ var _ = Describe("AWS Lambda", func() {
 		Expect(err).NotTo(HaveOccurred())
 	}
 
-	validateLambda := func(offset int, envoyPort uint32, substring string) {
+	type lambdaValidationParams struct {
```
Comment (sam-heilbron) [inline test/e2e/aws_test.go:90]: purely a nit, feel free to ignore. Given that we using the more compact/expressive custom HttpResponseMatcher, I find this indirection (using a utilitiy method with params) takes away some of the clarity of the inputs / expected outputs of each test.

Since certain tests require only a subset of these values, we have to do a lot of default setting.

I would migrate this duplicate code to just have each test do:

```
Eventually:
- Build request
- Execute request
- Assert response matches values
```

**Review by sam-heilbron — COMMENTED** [review]
I think this is a great improvement to the matcher library. Really only nits taht you can feel free to ignore

File: test/e2e/aws_test.go (hunk 19 above)
Comment (jbohanon) [reply test/e2e/aws_test.go:90]: Personally, I think passing in a params object makes the helper function much easier to reason about since the parameters must be named when they're defined in the object. I agree that the helper function is less than ideal, but I think that would be better tackled as part of https://github.com/solo-io/gloo/issues/7970

File: test/gomega/matchers/http.go (hunk 17 above)
Comment (jbohanon) [reply test/gomega/matchers/http.go]: changed the name and added a comment [84fb899](https://github.com/solo-io/gloo/pull/7854/commits/84fb8994f8c894109760d91f941377aae85938c5)

File: test/e2e/aws_test.go (hunk 19 above)
Comment (sam-heilbron) [reply test/e2e/aws_test.go:90]: Totally fair!

**Review by sam-heilbron — APPROVED** [review]
Awesome work!

**Review by ben-taussig-solo — APPROVED** [review]
